server/mqtt.go

```diff
@@ -4476,33 +4570,43 @@ func (sess *mqttSession) ensurePubRelConsumerSubscription(c *client) error {
 		maxAckPending = mqttDefaultMaxAckPending
 	}
 
+	sess.mu.Lock()
+	pubRelSubscribed := sess.pubRelSubscribed
+	pubRelSubject := sess.pubRelSubject
+	pubRelDeliverySubjectB := sess.pubRelDeliverySubjectB
+	pubRelDeliverySubject := sess.pubRelDeliverySubject
+	pubRelConsumer := sess.pubRelConsumer
+	tmaxack := sess.tmaxack
+	idHash := sess.idHash
+	sess.mu.Unlock()
+
 	// Subscribe before the consumer is created so we don't loose any messages.
-	if !sess.pubRelSubscribed {
-		_, err := c.processSub(sess.pubRelDeliverySubjectB, nil, sess.pubRelDeliverySubjectB,
+	if !pubRelSubscribed {
+		_, err := c.processSub(pubRelDeliverySubjectB, nil, pubRelDeliverySubjectB,
 			mqttDeliverPubRelCb, false)
 		if err != nil {
-			c.Errorf("Unable to create subscription for JetStream consumer on %q: %v", sess.pubRelDeliverySubject, err)
+			c.Errorf("Unable to create subscription for JetStream consumer on %q: %v", pubRelDeliverySubject, err)
 			return err
 		}
-		sess.pubRelSubscribed = true
+		pubRelSubscribed = true
 	}
 
 	// Create the consumer if needed.
-	if sess.pubRelConsumer == nil {
+	if pubRelConsumer == nil {
```

Comment:
Need to handle the error that the consumer may already exost

Reply:
@kozlovic I do not seem to be getting no error when creating a consumer twice with identical configs. `JSConsumerAlreadyExists` is only returned if the config differs. Does it mean that only one gets created for the unique config, and I don't need to worry about this issue?

Reply:
It should be fine. Moreover, you are already getting the sess.pubRelConsumer when restoring a session. If the JS consumer was already created, you won't try to create again (you simply recreate the NATS subscription that receives messages from the JS consumer). 

The only think that I am noticing now, is that technically maybe we should do the tmaxack+maxAckPending check in the `if !pubRelSubscribed {` check because although we may not need to recreate the JS consumer, but we do the same in processJSConsumer(), so I think it is fine living like this.